src/domain/flush/flush.repository.ts

```diff
@@ -0,0 +1,69 @@
+import {
+  Inject,
+  Injectable,
+  UnprocessableEntityException,
+} from '@nestjs/common';
+import { CacheRouter } from '../../datasources/cache/cache.router';
+import {
+  CacheService,
+  ICacheService,
+} from '../../datasources/cache/cache.service.interface';
+import { IConfigApi } from '../interfaces/config-api.interface';
+import { InvalidationPatternDto } from './entities/invalidation-pattern.dto.entity';
+import { InvalidationTarget } from './entities/invalidation-target.entity';
+import { IFlushRepository } from './flush.repository.interface';
+
+@Injectable()
+export class FlushRepository implements IFlushRepository {
+  constructor(
+    @Inject(CacheService) private readonly cacheService: ICacheService,
+    @Inject(IConfigApi) private readonly configApi: IConfigApi,
+  ) {}
+
+  async execute(pattern: InvalidationPatternDto): Promise<void> {
+    switch (pattern.invalidate) {
+      case InvalidationTarget.Chains:
+        return this.invalidateChains();
+      case InvalidationTarget.Contracts:
+        return this.invalidateContracts();
+      case InvalidationTarget.Tokens:
+        return this.invalidateTokens(pattern);
+    }
+  }
+
+  private async invalidateChains(): Promise<void> {
+    const chains = await this.configApi.getChains();
+    await this.cacheService.delete(CacheRouter.getChainsCacheDir());
+    await Promise.all(
+      chains.results.map((chain) =>
+        this.cacheService.delete(CacheRouter.getChainCacheDir(chain.chainId)),
+      ),
+    );
+  }
+
+  private async invalidateContracts(): Promise<void> {
+    const chains = await this.configApi.getChains();
+    await Promise.all(
+      chains.results.map((chain) =>
+        this.cacheService.delete(
+          CacheRouter.getContractCacheDir(chain.chainId, ''),
+        ),
+      ),
+    );
+  }
+
+  private async invalidateTokens(
+    pattern: InvalidationPatternDto,
+  ): Promise<void> {
+    const chainId = pattern?.patternDetails?.chainId;
+    if (!chainId) {
+      throw new UnprocessableEntityException(
+        `Unprocessable cache invalidation pattern detail: ${JSON.stringify(
+          pattern.patternDetails,
+        )}`,
+      );
+    }
+    await this.cacheService.delete(CacheRouter.getTokensCacheDir(chainId));
+    await this.cacheService.delete(CacheRouter.getTokenCacheDir(chainId, ''));
```

Comment:
> Yes, I think this is derived from passing a `CacheDir` to the delete method (see [this discussion](https://github.com/5afe/safe-client-gateway-nest/pull/285#discussion_r1159723868)), maybe just passing a string representing the key instead of a full `CacheDir` makes more sense after all. This is also related with [this comment](https://github.com/5afe/safe-client-gateway-nest/pull/288#discussion_r1163904212).
> 
> My suggestion here is to just pass the key part to the `delete` function (because we are always dealing by key and ignoring the field part) and rename it: `ICacheService.deleteByKey(key: string): Promise<void>`, what do you think?

For now I think this is ok but I still think we should improve on this solution. if `delete` receives a key then we leak the implementation of what `CacheDir` contains. This is more prominent when you have `set(CacheDir)` and `get(CacheDIr)`.

But I would just move forward with your suggestion since we won't be tackling this in this PR 👍 Just providing food for thought now!

Reply:
I created [this issue](https://github.com/5afe/safe-client-gateway-nest/issues/297) to tackle those concerns, if you're OK with it, I'll work on a clearer solution there so we can move forward with this PR 🙂

Reply:
> I created [this issue](https://github.com/5afe/safe-client-gateway-nest/issues/297) to tackle those concerns, if you're OK with it, I'll work on a clearer solution there so we can move forward with this PR 🙂

Totally ok with tackling in a separate PR/discussion! 👍